File: LeetCode/0938. Range Sum of BST/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
41 / 41 test cases passed.
Runtime: 204 ms
Memory Usage: 23 MB
"""
class Solution:
    def rangeSumBST(self, root: TreeNode, low: int, high: int) -> int:
        if not root:
            return 0
        if low <= root.val <= high:
            return root.val + self.rangeSumBST(root.left, low, high) + self.rangeSumBST(root.right, low, high)
        return self.rangeSumBST(root.left, low, high) if root.val > low else self.rangeSumBST(root.right, low, high)

File: LeetCode/0938. Range Sum of BST/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {
 */

/**
 * 41 / 41 test cases passed.
 * Runtime: 180 ms
 * Memory Usage: 63.1 MB 
 */
class Solution {
public:
    void inorder(TreeNode* root, int low, int high, int* ans) {
        if (!root) return;
        inorder(root->left, low, high, ans);
        if (root->val > high) return;
        if (root->val >= low) *ans += root->val;
        inorder(root->right, low, high, ans);
    }
    int rangeSumBST(TreeNode* root, int low, int high) {
        int ans = 0;
        inorder(root, low, high, &ans);
        return ans;
    }
};
